gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAChallengeEndpoint.java

```diff
@@ -300,17 +312,19 @@ private Handler<AsyncResult<Void>> verifyHandler(RoutingContext routingContext,
                                                      FactorContext factorContext) {
         return h -> {
             if (h.failed()) {
+                String failureReason = isEnrolling(routingContext, factorProvider, factorContext) ? MFA_ENROLL_FAILED : MFA_CHALLENGE_FAILED;
+                updateAuditLog(failureReason, endUser, client, factor, factorContext, h.cause());
                 handleException(routingContext, ERROR_PARAM_KEY, "mfa_challenge_failed");
                 return;
             }
 
             if (factor.is(FIDO2)) {
-                handleFido2Factor(routingContext, client, endUser, code, h);
+                handleFido2Factor(routingContext, client, endUser, code, factor, factorContext, h);
                 return;
             }
             // save enrolled factor if needed and redirect to the original url
             routingContext.session().put(ConstantKeys.MFA_FACTOR_ID_CONTEXT_KEY, factorId);
-            if (routingContext.session().get(ConstantKeys.ENROLLED_FACTOR_ID_KEY) != null || factorProvider.useVariableFactorSecurity(factorContext)) {
+            if (isEnrolling(routingContext, factorProvider, factorContext)) {
                 enrolledFactor.setStatus(FactorStatus.ACTIVATED);
                 saveFactor(endUser, factorProvider.changeVariableFactorSecurity(enrolledFactor), fh -> {
                     if (fh.failed()) {
```

Comment:
**question**: Do we also have to log a Failure here :thinking:  ?

Reply:
done